### PR TITLE
Move Firestore to Auth interop library.

### PR DESCRIPTION
**Pull request by ryanwilson**
This will be one less dependency on Core's Auth changes and handle the instance lifecycle for deleted `FIRApp`s.

### CODE REVIEW

**Comment by paulb777** [comment]
You probably need cmake changes for FirebaseAuthInterop similar to the ones I made for GoogleUtilities.

**Comment by wilhuff** [comment]
I love continuous integration.

File: Firestore/Source/API/FIRFirestore.mm [hunk 1]
```diff
@@ -16,8 +16,13 @@
 
 #import "FIRFirestore.h"
 
+#import <FirebaseAuthInterop/FIRAuthInterop.h>
```
Comment (wilhuff) [inline Firestore/Source/API/FIRFirestore.mm]: This file is now getting huge. Is there any way we can split out these other things into separate parts?

Ideally you should be able to reason about FIRFirestore without thinking about interop stuff and be able to reason about the interop stuff separately too.

File: Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.mm [hunk 2]
```diff
@@ -31,8 +33,9 @@
 namespace firestore {
 namespace auth {
 
-FirebaseCredentialsProvider::FirebaseCredentialsProvider(FIRApp* app)
-    : contents_(std::make_shared<Contents>(app, User::FromUid([app getUID]))) {
+FirebaseCredentialsProvider::FirebaseCredentialsProvider(FIRApp* app) {
+  id<FIRAuthInterop> auth = FIR_COMPONENT(FIRAuthInterop, app.container);
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.mm]: The amount of boilerplate in the tests here makes me think that we should pull the interaction with the container out of this class.

As a general principle we should isolate code from the dependency injector and I think this principle should apply all the way out to FIRFirestore. By the time you've constructed a FIRFirestore instance we should be able to supply it with all the dependencies it needs. 

Please change this constructor to take `FIRApp* app, id<FIRAuthInterop> auth`.

File: Firestore/Source/API/FIRFirestore.mm (hunk 1 above)
Comment (ryanwilson) [reply Firestore/Source/API/FIRFirestore.mm]: Chatted offline, but adding it here for the record: moving the interop code to a new class called `FSTFirestoreComponent` that will handle registration and instance creation/retrieval.

**Comment by wilhuff** [comment]
Once I clear up the build problems, the test fails post run here:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x20)
  * frame #0: 0x00007fff78d9ad6f libobjc.A.dylib`objc_retain + 31
    frame #1: 0x000000010008e23a firebase_firestore_auth_apple_test`-[FIRComponentContainer instanceForProtocol:](self=0x0000000100415ff0, _cmd="instanceForProtocol:", protocol=0x00000001000b8690) at FIRComponentContainer.m:172
    frame #2: 0x000000010008e8f3 firebase_firestore_auth_apple_test`+[FIRComponentType instanceForProtocol:inContainer:](self=FIRComponentType, _cmd="instanceForProtocol:inContainer:", protocol=0x00000001000b8690, container=0x0000000100415ff0) at FIRComponentType.m:25
    frame #3: 0x0000000100010dd0 firebase_firestore_auth_apple_test`firebase::firestore::auth::FirebaseCredentialsProvider::GetToken(this=0x00007ffeefbfef90, completion=0x00007ffeefbfef60)>) at firebase_credentials_provider_apple.mm:127
    frame #4: 0x00000001000026a1 firebase_firestore_auth_apple_test`firebase::firestore::auth::FirebaseCredentialsProviderTest_GetTokenUnauthenticated_Test::TestBody(this=0x0000000100406e20) at firebase_credentials_provider_test.mm:116
    frame #5: 0x000000010007697e firebase_firestore_auth_apple_test`void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(object=0x0000000100406e20, method=21 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="the test body")(), char const*) at gtest.cc:2402
    frame #6: 0x00000001000303eb firebase_firestore_auth_apple_test`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(object=0x0000000100406e20, method=21 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="the test body")(), char const*) at gtest.cc:2438
    frame #7: 0x0000000100030316 firebase_firestore_auth_apple_test`testing::Test::Run(this=0x0000000100406e20) at gtest.cc:2474
```

**Comment by ryanwilson** [comment]
@wilhuff I believe I addressed everything now and travis is passing, thanks again for the help with CMake!

File: Firestore/Source/API/FSTFirestoreComponent.mm [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    FSTThrowInvalidArgument(@"database identifier may not be nil.");
+  }
+
+  NSString *key = [NSString stringWithFormat:@"%@|%@", self.app.name, database];
+
+  // Get the component from the container.
+  @synchronized(self.instances) {
+    FIRFirestore *firestore = _instances[key];
+    if (!firestore) {
+      // Question for Gil: Do we want to key based on app name as well here?
```
Comment (wilhuff) [inline Firestore/Source/API/FSTFirestoreComponent.mm]: I'm not opposed, but I think the default app name is fugly, so I'd prefer the default to remain as is. Are you thinking of something like this?

```c++
std::string queue_name{"com.google.firebase.firestore"};
if (self.app.name is not the default app) {
  absl::StrAppend(&queue_name, ".", util::MakeStringView(self.app.name));
}
... dispatch_queue_create(queue_name.c_str(), DISPATCH_QUEUE_SERIAL);
```

File: Firestore/core/test/firebase/firestore/auth/firebase_credentials_provider_test.mm [hunk 4]
```diff
@@ -27,30 +29,64 @@
 
 #include "gtest/gtest.h"
 
-namespace firebase {
-namespace firestore {
-namespace auth {
+/// Testing interface for ComponentContainer (required for Auth).
+@interface FIRComponentContainer ()
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/auth/firebase_credentials_provider_test.mm]: This is now dead code, no?

Comment (ryanwilson) [reply Firestore/core/test/firebase/firestore/auth/firebase_credentials_provider_test.mm]: Nice catch, will remove.

File: Firestore/Source/API/FSTFirestoreComponent.mm (hunk 3 above)
Comment (ryanwilson) [reply Firestore/Source/API/FSTFirestoreComponent.mm]: That sounds good to me, do you foresee any issues with instances being on separate `dispatch_queue`s?

Comment (wilhuff) [reply Firestore/Source/API/FSTFirestoreComponent.mm]: Nope. The only thing we care about is that within an instance all operations happen on the same dispatch queue.

**Pull request merged**